### DISCUSSION TITLE
How to add all-capsules/container listener or observer for logging

### DISCUSSION THREAD

**Discussion by Kypsis** [GregoryConrad/rearch-dart#107] [Q&A] [answered]
How to add all-capsules/container listener for app-wide capsule state logging or some sort of Riverpod style ProviderObserver to container? The following part that probably deals with this in documentation is empty: https://rearch.gsconrad.com/paradigms/listeners

**Comment by GregoryConrad** [comment, marked as answer]
@Kypsis Yea, I never got around to actually finishing that page. Maybe I can tackle it this evening.

The idea is to just create a new capsule and opt-in to the capsules you want to listen to, knowing that they will never be freed from the container when you use listeners. Normally, ReArch will garbage-collect any capsules that don't use side effects when their dependencies change, but if you are adding a listener to those capsules, then they won't be freed. This doesn't matter at all if you are listening to a stateful capsule, but it can matter if you have a series of capsules that just do some simple data transformations that should be freed when no longer in use.

Here's a quick example:

```dart
(int, void Function(int)) aCapsule(CapsuleHandle use) => use.state(0);
void aListener(CapsuleHandle use) {
  use.asListener(); // mark this capsule as a listener (so it won't be automatically garbage-collected & not receive updates
  print(use(aCapsule)); // listen to whatever you want
}
final container = CapsuleContainer();
container.read(aListener); // instantiate the listener (which will call it once)
```

**Reply by GregoryConrad** [reply]
To just log the data of a list of capsules:
```dart
// in your listener capsule
use.asListener();
[
  aCapsule,
  bCapsule,
  // ...
].map(use).forEach(myLoggingFunction);
```

**Reply by Kypsis** [reply]
This works, thanks! Although I used
```dart 
if (kDebugMode) CapsuleContainerProvider.containerOf(context).read(loggerCapsule);
```
to get the container from the bootstrapper.

**Reply by GregoryConrad** [reply]
@Kypsis I finally wrote the documentation page in question, https://rearch.gsconrad.com/paradigms/listeners

If you get a chance, I'd appreciate your input (is anything unclear)?

**Reply by Kypsis** [reply]
The article is very well written, clear and describes the gotchas. So a big 👍 from me.
